clang/include/clang/AST/Type.h

```diff
@@ -4940,6 +4955,78 @@ class FunctionEffectsRef {
   void dump(llvm::raw_ostream &OS) const;
 };
 
+/// A mutable set of FunctionEffect::Kind.
+class FunctionEffectKindSet {
+  // For now this only needs to be a bitmap.
+  constexpr static size_t EndBitPos = 8;
+  using KindBitsT = std::bitset<EndBitPos>;
+
+  KindBitsT KindBits{};
+
+  explicit FunctionEffectKindSet(KindBitsT KB) : KindBits(KB) {}
+
+  constexpr static size_t kindToPos(FunctionEffect::Kind K) {
+    return static_cast<size_t>(K);
+  }
+
+public:
+  FunctionEffectKindSet() = default;
+  explicit FunctionEffectKindSet(FunctionEffectsRef FX) { insert(FX); }
+
+  // Iterates through the bits which are set.
+  class iterator {
```

Comment:
Wow, it still builds when private, but I don't understand. Apparently the `for` loops which walk through this thing aren't being foiled by the privateness of the iterator.

Reply:
As I understand it, you can still return values of inaccessible type just fine from functions. You just can’t *name* `iterator` outside of this class, so `auto x = set.begin()` works, but `FunctionEffectKindSet::iterator x = set.begin()` doesn’t, and since range-based `for` loops are specified to use `auto`/`auto&&`, everything works fine.

Reply:
TIL. Thanks.